ext/spl/spl_iterators.c

```diff
@@ -3171,12 +3171,43 @@ PHP_FUNCTION(iterator_to_array)
 	zval  *obj;
 	bool use_keys = 1;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "O|b", &obj, zend_ce_traversable, &use_keys) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_ITERABLE(obj)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_BOOL(use_keys)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (Z_TYPE_P(obj) == IS_ARRAY) {
+		if (use_keys) {
+			RETURN_COPY(obj);
+		}
+		else {
+			zend_array *arrval;
+			zend_long arrlen;
+			zval *entry;
+
+			arrval = Z_ARRVAL_P(obj);
+
+			arrlen = zend_hash_num_elements(arrval);
 
-	array_init(return_value);
-	spl_iterator_apply(obj, use_keys ? spl_iterator_to_array_apply : spl_iterator_to_values_apply, (void*)return_value);
+			array_init_size(return_value, arrlen);
+			zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
+
+			ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
+				ZEND_HASH_FOREACH_VAL(arrval, entry) {
+					if (UNEXPECTED(Z_ISREF_P(entry) && Z_REFCOUNT_P(entry) == 1)) {
+						entry = Z_REFVAL_P(entry);
+					}
+					Z_TRY_ADDREF_P(entry);
+					ZEND_HASH_FILL_ADD(entry);
+				} ZEND_HASH_FOREACH_END();
+			} ZEND_HASH_FILL_END();
+		}
```

Comment:
This is copied from `\array_values()`. Is there a simpler solution for this or is it possible to call `\array_values()` internally?

Reply:
Probably yes

Reply:
Okay, thanks. Done.